### PR TITLE
Remove realtime segment metrics if it's destroyed

### PR DESCRIPTION
**Pull request by jackjlli**
## Description
This PR adds the logic to remove the realtime segment metrics if it gets destroyed.
This step is essential as the instance may not be the target location for some of the partitions.

 E.g. if the number of partitions increases, or a host swap is needed, realtime segment assignment needs to be re-calculated. The target location for some partitions may change, and the current host remains to run. In this case, the current server would still keep the states of the old partitions, which no longer resides in this host any more, thus causes false positive information to the metric system. 

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/AbstractMetrics.java [hunk 1]
```diff
@@ -462,16 +463,48 @@ public void addCallbackGaugeIfNeeded(final String metricName, final Callable<Lon
...
-        }));
+    PinotMetricUtils
+        .makeGauge(_metricsRegistry, PinotMetricUtils.makePinotMetricName(_clazz, _metricPrefix + metricName),
+            PinotMetricUtils.makePinotGauge(avoid -> {
+              try {
+                return valueCallback.call();
+              } catch (Exception e) {
+                LOGGER.error("Caught exception", e);
+                Utils.rethrowException(e);
+                throw new AssertionError("Should not reach this");
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/metrics/AbstractMetrics.java:474]: I assume this was added when testing Utils.rethrowException(). Do we need this assertion here?

**Review by mcvsubbu — COMMENTED** [review]
1. I assume YammerMetrics works correctly when metrics are added or remove at the same time. It is possible that the destroy()  method is called even as a new consuming segment gets added to a host.
2. You should add a comment in javadocs `PinotMetricsRegistry` interface that it should be thread-safe against adds and removes of the same metric.
3. Add a comment here as well, saying we expect the add/remove to work correctly in  case of being invoked across multiple threads.

**Comment by jackjlli** [comment]
> 1. I assume YammerMetrics works correctly when metrics are added or remove at the same time. It is possible that the destroy()  method is called even as a new consuming segment gets added to a host.
> 2. You should add a comment in javadocs `PinotMetricsRegistry` interface that it should be thread-safe against adds and removes of the same metric.
> 3. Add a comment here as well, saying we expect the add/remove to work correctly in  case of being invoked across multiple threads.

1. Correct, yammer metrics works correctly on handling add and remove at the same time.
2.3. Comments added.

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/AbstractMetrics.java (hunk 1 above)
Comment (jackjlli) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/AbstractMetrics.java:474]: The `Utils.rethrowException(e);` is just a method name. Thus, it requires an extra statement as the final return statement. The assertion statement here is just to denote that this line won't be hit at any time. If not, we may have a bug in `Utils.rethrowException(e);` method.

**Review by mcvsubbu — APPROVED** [review]

**Pull request merged**